Profiler readme enhancements (#473)
signalfx/splunk-otel-java: profiler/README.md

> 🚨 The profiler is in beta and its use is currently restricted.

Reviewer: We usually add
```markdown
> :construction: &nbsp;Status: Experimental
```
to beta/experimental features' docs; see [metrics.md](https://github.com/signalfx/splunk-otel-java/blob/main/docs/metrics.md) for an example.

Author: Sounds good. 👍🏻

> 🚨 The profiler is in beta and its use is currently restricted.

Reviewer 2: What does it mean "restricted"?

Author: It means that users shouldn't have an expectation yet that it just works and that they need to be working directly with us in order to use the thing in a meaningful way.

Author: Anyway, it's been changed to experimental.

> 🚨 The profiler is in beta and its use is currently restricted.

Reviewer 3: ```suggestion
> **Warning** The profiler feature is experimental.
```

Author: Yeah, I'll go with that construction bit that @mateuszrzeszutek noted above.

> # About

Reviewer 3: ```suggestion
# About the Java profiler
```

> The `splunk-otel-java` agent includes an "always-on" profiler that can be enabled with a configuration
> setting. This profiler periodically captures the call stack state for all JVM threads and
> saves these to the Splunk Observability Cloud. Users can then view a flamegraph of application
> call stacks and inspect individual code-level call stacks for relevant traces.

Reviewer 3: ```suggestion
The `splunk-otel-java` agent includes a continuous profiler that can be enabled at any time. This profiler periodically captures the call stack state for all JVM threads and send them to Splunk Observability Cloud. You can then view a flamegraph of application call stacks and inspect individual code-level call stacks for relevant traces.
```

(DD calls that "Continuous profiler".)

Author: I don't mind the change around "always-on" to "continuous", no preference.

 I'm opposed to changing the wording to "enabled at any time" because it's in accurate. It has to be configured at startup time and cannot be changed in a running application, which is what the "enabled at any time" suggests.

Reviewer 3: Suggestion updated.

Author: still says "at any time" for me tho.

Reviewer 3: How about now?

Author: Now it says it is always enabled...which is not true. By default it is NOT enabled (yet).  :)

Reviewer 3: Gotcha. Reworded.

> ## How?

Reviewer 3: ```suggestion
## How does the profiler works?
```

> The profiler leverages the JVM's Java Flight Recorder (JFR) to perform periodic call

Reviewer 2: ```suggestion
The profiler leverages the JVM's JDK Flight Recorder (JFR) to perform periodic call
```

> stack sampling. Every recording period, a new JFR recording file is flushed to disk and then
> the events are replayed.

Reviewer 3: ```suggestion
stack sampling. For every recording period, a new JFR recording file is flushed to disk and events are replayed.
```

In order to associate `jdk.ThreadDump` events from JFR with OpenTelemetry spans, a custom

> OpenTelemetry `ContextStorage` implementation is used to synthesize `otel.ContextAttached`

Reviewer 2: ```suggestion
OpenTelemetry `ContextStorage` implementation is used to emit `otel.ContextAttached`
```

> In order to associate `jdk.ThreadDump` events from JFR with OpenTelemetry spans, a custom
> OpenTelemetry `ContextStorage` implementation is used to synthesize `otel.ContextAttached`
> events each time a span context change occurs. These context changes are then tracked
> during replay so that a call stack can be associated with the appropriate span.
> 
> Stack trace data is embedded as a string inside of an OTLP logs payload. The
> [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector)
> will detect this profiling data inside of OTLP logs and will help it along
> its ingest path.

Reviewer 3: ```suggestion
To associate `jdk.ThreadDump` events from JFR with OpenTelemetry spans, a custom
OpenTelemetry `ContextStorage` implementation generates `otel.ContextAttached`
events each time a span context change occurs. These context changes are then tracked
during replay so that a call stack can be associated with the appropriate span.

Stack trace data is embedded as a string inside of an OTLP logs payload. The
[Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) detects
this profiling data inside of OTLP logs and help it along its ingest path.
```

What do we mean by "help it along its ingest path" technically?

# Requirements

> * Java 8 above 8u262, Java 11+

Reviewer: Nit:
```suggestion
* Java 8 above 8u262, or Java 11+
```

> * Java 8 above 8u262, Java 11+

Reviewer 3: ```suggestion
* Java 8u262 or higher
```

Java 8 or higher should be enough, no? Java 9 and 10 are not supported.

> * 100MB free disk space
> * [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) > 0.33.1
> * The profiler is enabled at startup (off by default, see Configuration below)

Reviewer 3: ```suggestion
* 100 MB of free disk space
* [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) version 0.33.1 or higher
* Profiler is enabled at startup (disabled by default, see the Configuration section)
```

> # Configuration
> 
> The configuration items below are available. In order for the profiler to be enabled,
> you MUST pass `-Dsplunk.profiler.enabled=true` to your JVM or set the `SPLUNK_PROFILER_ENABLED`
> environment variable to your JVM process.
> 
> We strongly recommend using defaults for all other settings.
> 
> Like all other OpenTelemetry Java configurations, the following items can be used as a system
> property or as environment variable (by making it uppercase and replacing dots with underscores).
> 
> | name                                     | default                | description                               |
> |------------------------------------------|------------------------|-------------------------------------------|

Reviewer 3: ```suggestion
# Enable the profiler

To enable the profiler, pass the `-Dsplunk.profiler.enabled=true` argument to your JVM or set the `SPLUNK_PROFILER_ENABLED` environment variable to your JVM process.

# Configuration settings

Like all other OpenTelemetry Java configurations, the following settings can be used as a system
property or as environment variables (by making them uppercase and replacing dots with underscores).

> We strongly recommend using defaults for the following settings.

| Setting                                     | Default                | Description                               |
|------------------------------------------|------------------------|-------------------------------------------|
```

|`splunk.profiler.period.{eventName}`      | n/a                    | customize period (in ms) for a specific jfr event. For example, to set the ThreadDump frequency to 1s (1000ms): `-Dsplunk.profiler.period.threaddump=1000` |
|`splunk.profiler.tlab.enabled`            | false                  | set to `false` to disable TLAB memory events |
|`splunk.profiler.include.agent.internals` | false                  | set to `true` to include agent internal call stacks |

# Escape hatch

The profiler limits its own behavior under two conditions:

> * Free disk space is too low (< 100MB free))

Reviewer 2: ```suggestion
* Free disk space is too low (< 100MB free)
```

* More than 5 minutes worth of JFR files are backed up on disk

> If a recording is already running when this condition is encountered, it will
> be stopped. Later, when the conditions are no longer applicable, then recording will
> be restarted and the profiler will resume operation.

Reviewer 3: ```suggestion
If a recording is already running when this condition is met, the recording stops. When
the conditions are no longer applicable, the recording and the profiler resume operation.
```

# FAQ / Troubleshooting

> ### How do I know if it's working?
> 
> At startup, the agent will log the string "JFR profiler is active" at `INFO`. You can grep for this in your logs to see
> something like this:
> ```
> [otel.javaagent 2021-09-28 18:17:04:246 +0000] [main] INFO com.splunk.opentelemetry.profiler.JfrActivator - JFR profiler is active.
> ```

Reviewer 3: ```suggestion
### Check that the profiler is working

If the profiler is enabled, the agent logs the string "JFR profiler is active" as an `INFO` entry at startup. You can grep your logs to find entries similar to the following:
```
[otel.javaagent 2021-09-28 18:17:04:246 +0000] [main] INFO com.splunk.opentelemetry.profiler.JfrActivator - JFR profiler is active.
```
```

> ### How can I see the profiler configuration?
> 
> The agent will log the profiling configuration at `INFO` during startup. You can grep for the string
> `com.splunk.opentelemetry.profiler.ConfigurationLogger` to see something like this:
> 

Reviewer 3: ```suggestion
### Retrieve the profiler configuration settings

The agent logs the profiling configuration as `INFO` entries at startup. You can grep for the string
`com.splunk.opentelemetry.profiler.ConfigurationLogger` to see entries similar to the following:
```

Reviewer 3: Okay :D

Reviewer 3: Suggestion updated

[otel.javaagent 2021-09-28 18:17:04:237 +0000] [main] INFO <snip> - Profiler configuration:
[otel.javaagent 2021-09-28 18:17:04:238 +0000] [main] INFO <snip> -                 splunk.profiler.enabled : true
[otel.javaagent 2021-09-28 18:17:04:239 +0000] [main] INFO <snip> -               splunk.profiler.directory : .
[otel.javaagent 2021-09-28 18:17:04:244 +0000] [main] INFO <snip> -      splunk.profiler.recording.duration : 20s
[otel.javaagent 2021-09-28 18:17:04:244 +0000] [main] INFO <snip> -              splunk.profiler.keep-files : false
[otel.javaagent 2021-09-28 18:17:04:245 +0000] [main] INFO <snip> -           splunk.profiler.logs-endpoint : null
[otel.javaagent 2021-09-28 18:17:04:245 +0000] [main] INFO <snip> -             otel.exporter.otlp.endpoint : http://collector:4317
[otel.javaagent 2021-09-28 18:17:04:245 +0000] [main] INFO <snip> -            splunk.profiler.tlab.enabled : false
[otel.javaagent 2021-09-28 18:17:04:246 +0000] [main] INFO <snip> -   splunk.profiler.period.jdk.threaddump : null
[otel.javaagent 2021-09-28 18:17:04:246 +0000] [main] INFO <snip> - -----------------------
```

> ### What about this escape hatch?
> 
> If the escape hatch becomes active, it will log with `com.splunk.opentelemetry.profiler.RecordingEscapeHatch`
> (you can grep for this in the logs). You may also look for `"** THIS WILL RESULT IN LOSS OF PROFILING DATA **"`
> as a big hint that things are not well.
> 
> You may need to free up some disk space and/or give the JVM more resources.

Reviewer 3: ```suggestion
### Check if the escape hatch activated

To check whether the escape hatch activated, grep your logs for `com.splunk.opentelemetry.profiler.RecordingEscapeHatch` entries. 

You can also look
 for the `"** THIS WILL RESULT IN LOSS OF PROFILING DATA **"` warning. If you find that message, you might need to free up some disk space or provide more resources to the JVM.
```

Reviewer 3: Yeah! xD  

Original question is a bit too conversational, but let's roll with it. FAQ style :)

Author: > Yeah! xD
> 
> Original question is a bit too conversational, but let's roll with it. FAQ style :) Suggestion updated.

I still don't see an updated suggestion. :/

Reviewer 3: Now

### What if I'm on an unsupported JVM?

If your JVM does not support JFR, this is detected by the profiler at startup. It will log

> "Java Flight Recorder (JFR) is not available in this JVM. Profiling is disabled." at `WARN`,

Reviewer 2: The official name is "JDK Flight Recorder"

Author: https://github.com/signalfx/splunk-otel-java/issues/476

> ### What if I'm on an unsupported JVM?
> 
> If your JVM does not support JFR, this is detected by the profiler at startup. It will log
> "Java Flight Recorder (JFR) is not available in this JVM. Profiling is disabled." at `WARN`,
> and you can grep for this in the logs.

Reviewer 3: ```suggestion
### Using unsupported JVMs

If your JVM does not support JFR, the profiler emits a warning at startup with the message `Java Flight Recorder (JFR) is not available in this JVM. Profiling is disabled`.
```

Author: I like this change, but I don't want to lose sight of the fact that this is logged (emit is vague) to the logs and can be found by searching the logs. Going to try some hybrid and see what you think.

Author: Please take another look at this part @theletterf

Reviewer 3: Went for "logs a warning"

> ### Why is the OTLP/logs exporter complaining?
> 
> This can happen for a number of reasons, but is generally caused by the collector being misconfigured.
> If the logs cannot be exported to the collector, the profiling data will not be available in
> the UI.

Reviewer 3: ```suggestion
### Why is the OTLP/logs exporter complaining?

Collector configuration issues may prevent logs from being exported and profiling data from showing in Splunk Observability Cloud.
```

> Things to check:

Reviewer 3: ```suggestion
Check for the following common issues:
```

> * Look at the values of the agent's configuration: `splunk.profiler.logs-endpoint` and `otel.exporter.otlp.endpoint`. Hint: they are logged
> at startup (see above).

Reviewer 3: ```suggestion
* Check the agent configurations `splunk.profiler.logs-endpoint` and `otel.exporter.otlp.endpoint`, which are logged at startup.
```

> * Verify that a collector is actually running at that endpoint and that the
> application host/container can resolve any hostnames and actually connect to the given OTLP port (default: 4317)

Reviewer 3: ```suggestion
* Verify that a collector is running at the designated endpoint and that the application host or container can resolve any hostnames and connect to the given OTLP port (default is `4317`).
```

> * Make sure you are running the [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector)
> and that the version is 0.33.1 or greater. Other collector distributions may not be able to route the log
> data containing profiles correctly.

Reviewer 3: ```suggestion
* Make sure you are running the [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) and that the version is 0.33.1 or higher. Other collector distributions may not be able to route the log data containing profiles.
```

* Make sure that the collector is configured correctly to handle profiling data. By default, the
[Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) handles this, but
a custom configuration might have overridden some settings. Make sure that an OTLP _receiver_ is configured in the collector
and that an exporter is configured for `splunk_hec` export. Ensure that the `token` and `endpoint` fields
are [correctly configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/splunkhecreceiver#configuration).

> Lastly, double check that the logs _pipeline_ is configured to use the OTLP receiver and `splunk_hec` exporter.

Reviewer 3: * Make sure that the collector is configured correctly to handle profiling data. By default, the [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) handles profiling data, but custom configuration might override profiling settings.
* * Make sure that an OTLP _receiver_ is configured in the collector and that an exporter is configured for `splunk_hec` export.
* * Make sure that the `token` and `endpoint` fields are [correctly configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/splunkhecreceiver#configuration).
* * Make sure that the logs _pipeline_ is configured to use the OTLP receiver and `splunk_hec` exporter.

> ### Why is the OTLP/logs exporter complaining?
> 
> This can happen for a number of reasons, but is generally caused by the collector being misconfigured.
> If the logs cannot be exported to the collector, the profiling data will not be available in
> the UI.
> 
> Things to check:
> 
> * Look at the values of the agent's configuration: `splunk.profiler.logs-endpoint` and `otel.exporter.otlp.endpoint`. Hint: they are logged
> at startup (see above).
> * Verify that a collector is actually running at that endpoint and that the
> application host/container can resolve any hostnames and actually connect to the given OTLP port (default: 4317)
> * Make sure you are running the [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector)
> and that the version is 0.33.1 or greater. Other collector distributions may not be able to route the log
> data containing profiles correctly.
> * Make sure that the collector is configured correctly to handle profiling data. By default, the
> [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) handles this, but
> a custom configuration might have overridden some settings. Make sure that an OTLP _receiver_ is configured in the collector
> and that an exporter is configured for `splunk_hec` export. Ensure that the `token` and `endpoint` fields
> are [correctly configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/splunkhecreceiver#configuration).
> Lastly, double check that the logs _pipeline_ is configured to use the OTLP receiver and `splunk_hec` exporter.
> 

Reviewer 3: ```suggestion
### Troubleshooting OTLP and logs exporter issues

Collector configuration issues may prevent logs from being exported and profiling data from showing in Splunk Observability Cloud.

Check for the following common issues:

* Check the agent configurations `splunk.profiler.logs-endpoint` and `otel.exporter.otlp.endpoint`, which are logged at startup.
* Verify that a collector is running at the designated endpoint and that the application host or container can resolve any hostnames and connect to the given OTLP port (default is `4317`).
* Make sure you are running the [Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) and that the version is 0.33.1 or higher. Other collector distributions may not be able to route the log
data containing profiles.
* Make sure that the collector is configured correctly to handle profiling data. By default, the
[Splunk OpenTelemetry Connector](https://github.com/signalfx/splunk-otel-collector) handles profiling data, but
a custom configuration might override profiling settings. Make sure that an OTLP _receiver_ is configured in the collector and that an exporter is configured for `splunk_hec` export. 
* Make sure that the `token` and `endpoint` fields are [correctly configured](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/splunkhecreceiver#configuration).
* Make sure that the logs _pipeline_ is configured to use the OTLP receiver and `splunk_hec` exporter.
```

Author: It's really hard to see what the suggested changes are in such a big multiline block. Can we make smaller suggestions?

Reviewer 3: You're right!

In this particular case, I'd keep the new title. Or we clarify the original question a bit.

Author: Can you please resolve this large conversation/suggestion and create smaller ones with 1-2 lines? I can't make sense of the suggested changes here...

> ### Can I tell the agent to ignore some threads?
> 
> Not yet. Some JVM internal threads are automatically omitted, but there is no user-serviceable mechanism
> for omitting or filtering threads. If you need this, please file an issue.

Reviewer 3: ```suggestion
### Can I tell the agent to ignore some threads?

Not yet. Some JVM internal threads are automatically omitted, but there is no user-serviceable mechanism
for omitting or filtering threads. If you need this, file an issue.
```
